### ISSUE TITLE
Super slow when opening files using v0.9.16

### ISSUE THREAD

**Issue by hongyuanjia** [jalvesaq/Nvim-R#602]
Opening files in the latest Nvim-R (v0.9.16) are really slow on Linux. It took 15s+ to open an [R script](https://github.com/hongyuanjia/eplusr/blob/master/R/impl-idf.R) with ~5000 lines. Profiling showed that `SourceRFunList()` is the function that causes the slowness. Nvim-R (v0.9.15) did not have this issue. 

```
FUNCTION  SourceRFunList()
    Defined: ~/.vim/plugged/Nvim-R/R/functions.vim:48
Called 30 times
Total time:  15.324936
 Self time:  15.299286

count  total (s)   self (s)
   30              0.000200     if exists("g:R_hi_fun") && g:R_hi_fun == 0
                                    return
   30              0.000021     endif
                            
   30              0.000162     let fnm = g:rplugin.compldir . '/fun_' . a:lib
                            
   30              0.000093     if has_key(g:rplugin, "localfun")
                                    call UpdateLocalFunctions(g:rplugin.localfun)
   30              0.000022     endif
                            
                                " Highlight R functions
   30              0.000113     if !exists("g:R_hi_fun_paren") || g:R_hi_fun_paren == 0
   30  15.323068  15.297418         exe "source " . substitute(fnm, ' ', '\\ ', 'g')
                                else
                                    let lines = readfile(fnm)
                                    for line in lines
                                        let newline = substitute(line, "\\.", "\\\\.", "g")
                                        if substitute(line, "syn keyword rFunction ", "", "") =~ "[ ']"
                                            let newline = substitute(newline, "keyword rFunction ", "match rSpaceFun /`\\\\zs", "")
                                            exe newline . "\\ze`\\s*(/ contained"
                                        else
                                            let newline = substitute(newline, "keyword rFunction ", "match rFunction /\\\\<", "")
                                            exe newline . "\\s*\\ze(/"
                                        endif
                                    endfor
   30              0.000026     endif
```

**Comment by jalvesaq** [comment]
The Vim scripts to highlight function names were being sourced multiple times. I changed the code, and now they are sourced only once, and only after the `nclientserver` is running. I would do this in a new branch but pushed the changes into the master branch by mistake.

The disadvantage of the new code is that function names are not highlighted at Vim's startup, only after one or two seconds after the startup (when the `nclientserver` starts).  The main advantage is a faster startup. The postponed highlighting is also useful to inform the user that the `nclientserver` is already running, that is, things such omni completion are ready to be used.

 Could you, please, update the plugin and tell me your opinion on the change?

**Comment by hongyuanjia** [comment]
Thanks for the info. I will pull the latest to see how it works.

**Comment by jalvesaq** [comment]
Vim's syntax highlighting is slow, and R loads thousands of functions at startup. It will be faster if functions are not highlighted individually.

If you put in your vimrc/init.vim

```vim
let r_syntax_fun_pattern = 1
```
then, all words followed by an opening parenthesis will be highlighted as functions, and the lists of functions loaded in R's memory will not be used at all. That is, Vim will not try to match each keyword with thousands of patterns.

**Comment by jalvesaq** [comment]
I'll close the issue because further improvement requires asynchronous syntax highlighting, which is not possible in Neovim yet.
